Use ipyfilechooser for file dialog

astrofrog opened the pull request:
This is an experiment with using ipyfilechooser as the file dialog, which allows us to select file paths on disk as opposed to having to upload the data as #186 does. This does mean the file browser is not 'native' but instead implemented using ipywidgets, so it is not as pretty, but on the other hand it is better than passing a file path! Here is a preview:

![Peek 2021-04-27 17-05](https://user-images.githubusercontent.com/314716/116274800-e719ab80-a77a-11eb-9d43-79ec1d50c25f.gif)

I had to hack around and use private API on ipyfilechooser to hide some of their buttons so this should be treated as experimental. If people like this I can work on trying to expose what we need in a public API in ipyfilechooser or in the worst case forking it and bundling the modified version here (it is pretty simple code)

This should otherwise be functional so feel free to try it out!

Fix #156, fix #546

eteq commented:
Note this addresses #546  (although note https://github.com/spacetelescope/jdaviz/issues/546#issuecomment-832004947 which is a bit larger in scope)

astrofrog commented:
@eteq - should I go ahead and finalize this PR to avoid private API calls? Or will we wait for anyone else to weigh in?

pllim commented:
I am not convinced that this is not introducing another single-point-of-failure into the stack. This package is maintained by one person that is not involved in our org...

astrofrog commented:
To be clear since it is a very small package and I would need to modify it I would fork it and bundle the forked version in jdaviz so we would not add a dependency.

pllim commented:
If you add it as bundle, please remember to copy over its license. Thanks! (It's not GPL, phew.)

javerbukh commented:
This works great! I think if you do what you mentioned in your comment, we can take this out of draft form and I'll approve it.

astrofrog commented:
@javerbukh @pllim - I've now refactored this to include a bundled forked and simplified version of ipyfilechooser. I could probably simplify the file_chooser.py code even further but I thought it might not be worth the effort if this is going to just be a temporary solution for now.

This is now ready for review and for trying out!

pllim reviewed (commented):
Maybe we want to also test this on Windows, I volunteer:

* @duytnguyendtn -- Native Windows
* @eteq -- Windows with WSL2

I'll report back from Ancient Linux in a min...

pllim commented on jdaviz/configs/default/plugins/data_tools/file_chooser.py:
```diff
@@ -0,0 +1,443 @@
+# The code in this file has been forked from the ipyfilechooser package which
+# was originally released under the following license:
+#
+# MIT License
```
Oh ops... this is now resolved.

pllim reviewed (approved):
On RHEL 7 + Chrome:

* Dialog pops up but from `HOME`, not working directory, as I would expect. Not a deal breaker, but also very inconvenient.
* Folder unicode does not render for me. Do we really need unicode? I think a trailing slash is indication enough that the entry is a folder? Also not a deal breaker.
* The single vs double click that Ricky observed got me too. I was double-clicking from muscle memory and then it started to act weird. Might want to document this somewhere.

I think this is a good first pass, but might want follow-up ticket(s) for future improvements.

astrofrog commented:
@pllim - I changed it so it starts from the current directory and also removed the icons in favor of a trailing slash.

I also would prefer double click events, I can investigate this but I think it would be worth looking into as a follow up PR to not hold this back. Folder selection requested by @duytnguyendtn requires this as then single click would just select a folder without opening it.

pllim commented:
> starts from the current directory

Ops... I didn't realize this would cause the "standalone" `viola` app to behave like this. Maybe a follow-up issue... I see this when I run `jdaviz --layout=imviz <filename>` from working directory, but somehow the dialog ends up in `/tmp/...` folder.

![Screenshot 2021-05-11 150328](https://user-images.githubusercontent.com/2090236/117870659-64a8e580-b26a-11eb-83a9-8cbc45fec24a.jpg)

astrofrog commented:
Rebased and fixed the voila issue @pllim mentioned (with the temporary directory)

pllim commented on jdaviz/configs/default/plugins/data_tools/data_tools.py:
```diff
@@ -11,22 +12,42 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if 'JDAVIZ_START_DIR' in os.environ:
+            start_path = os.environ['JDAVIZ_START_DIR']
+        else:
+            start_path = os.path.abspath(os.path.curdir)
```
Could be simplified? Though if you really want, you can also check if the env var is pointing to a real dir, but I am not sure if we have to.

```suggestion
        start_path = os.environ.get('JDAVIZ_START_DIR', os.path.curdir)
```

astrofrog replied:
Thanks!

pllim commented:
Small suggestion at astrofrog/jdaviz#3 , otherwise LGTM. Thanks!

astrofrog commented:
Rebased to get rid of the RTD failure

pllim commented:
Multiple approvals, so I took the liberty to merge. Thanks!
